File: cloudshell-temp/shared/contracts/alerts.ts
import { z } from "zod";

/** Timeframes like 'same','1m','5m','15m','1h','1d' */
export const Timeframe = z.string().min(2);

export const CompareWith = z.enum(["value", "indicator_component", "price"]);
export const Operator = z.enum([
  ">", "<", ">=", "<=", "equals", "crosses_above", "crosses_below",
]);

export const IndicatorRef = z.object({
  indicator: z.string(),           // e.g. "RSI", "EMA", "MACD"
  component: z.string().optional(),// e.g. "RSI", "Signal", "EMA"
  settings: z.record(z.any()).optional(),
});

export const Condition = z.object({
  id: z.string(),
  type: z.enum(["indicator", "price", "volume"]),
  // LHS
  indicator: z.string().optional(),         // when type='indicator'
  component: z.string().optional(),
  settings: z.record(z.any()).optional(),
  // Comparison
  operator: Operator,
  compareWith: CompareWith,
  compareValue: z.number().optional(),      // when compareWith='value'
  rhs: IndicatorRef.optional(),             // when compareWith='indicator_component'
  // Context
  timeframe: z.string().default("same"),
});

export const ConditionArray = z.array(Condition).min(1);
export const Logic = z.enum(["AND", "OR"]);

export const Action = z.discriminatedUnion("type", [
  z.object({ type: z.literal("notify") }),
  z.object({ type: z.literal("webhook"), url: z.string().url() }),
  // future
  z.object({ type: z.literal("bot"), bot_id: z.string() }),
]);

export const FireMode = z.enum(["per_bar", "per_close", "per_tick"]);

export const AlertCreate = z.object({
  symbol: z.string(),
  base_timeframe: Timeframe,
  conditions: ConditionArray,
  logic: Logic.default("AND"),
  action: Action.default({ type: "notify" }),
  status: z.enum(["active", "paused"]).default("active"),
  fireMode: FireMode.default("per_bar"),
});

export type TCondition = z.infer<typeof Condition>;
export type TAction = z.infer<typeof Action>;
export type TFireMode = z.infer<typeof FireMode>;
export type TAlertCreate = z.infer<typeof AlertCreate>;

File: cloudshell-temp/backend/shared/contracts/alerts.py
from dataclasses import dataclass
from typing import Optional, Dict, Any, List, Literal

Operator = Literal[">","<",">=","<=","equals","crosses_above","crosses_below"]
CompareWith = Literal["value","indicator_component","price"]

@dataclass
class IndicatorRef:
    indicator: str
    component: Optional[str] = None
    settings: Optional[Dict[str, Any]] = None

@dataclass
class Condition:
    id: str
    type: Literal["indicator","price","volume"]
    operator: Operator
    compareWith: CompareWith
    compareValue: Optional[float] = None
    rhs: Optional[IndicatorRef] = None
    indicator: Optional[str] = None
    component: Optional[str] = None
    settings: Optional[Dict[str, Any]] = None
    timeframe: str = "same"

@dataclass
class ActionNotify:
    type: Literal["notify"] = "notify"

@dataclass
class ActionWebhook:
    type: Literal["webhook"] = "webhook"
    url: str = ""

Action = ActionNotify | ActionWebhook  # extend later

@dataclass
class AlertCreate:
    symbol: str
    base_timeframe: str
    conditions: List[Condition]
    logic: Literal["AND","OR"] = "AND"
    action: Any = None
    status: Literal["active","paused"] = "active"



